Modularization of Ability

carlosgarciaq opened the pull request:
Before this patch, there were two 'Ability' classes: Ability and
SocialStream::Ability. Only 'SocialStream::Ability' was being used
for authorization by CanCan, though the one that was normally
superseded was 'Ability'.

After this patch, there is only one 'Ability' class, in app/models, that
includes the module SocialStream::Models::Ability.

The events module has also been modified to overload
SocialStream::Models::Ability with its rules.

Also, the generator social_stream:add_authorization_rules has been
added to easily supersede app/models/ability.rb
Thus, the user does not need to know that Ability should include
SocialStream::Ability and call the parent initializer when
overloading this class.

cgarciaq committed:
Modularization of Ability.

Before this patch, there were two 'Ability' classes: Ability and
SocialStream::Ability. Only 'SocialStream::Ability' was being used
for authorization by CanCan, though the one that was normally
superseded was 'Ability'.

After this patch, there is only one 'Ability' class, in app/models,
that includes the module SocialStream::Ability.

The events module has also been modified to overload
SocialStream::Ability with its rules.

Also, the generator social_stream:install has been
modified to create app/models/ability.rb
Thus, the user does not need to know that Ability should include
SocialStream::Ability and call the parent initializer when
overloading this class).
This generator detects if ability.rb exists and in this case
modifies it to include 'SocialStream::Ability'

atd commented:
Very good job @carlosgarciaq! The idea of the generator seems pretty helpful to me. I have only a couple of minor issues:
- What about preserving SocialStream::Base::Ability along with SocialStream::Ability? The idea is that the code in SocialStream::Base::Ability remains at the botton of the ancestors chain, so other social stream's plugins are able to overwrite settings in base.
- I would move the generator to social_stream-base, so it is also present to developers not using the full social_stream suite.

What do you think?

atd commented:
BTW, please remove the version related changes. There is already a release script in release.thor that updates them automagically

carlosgarciaq commented:
Hi @atd !
- No problem about the version files.
- I had the generator originally in social_stream-base, but decided that having it in social_stream was easier to find for the new user. I think keeping both would be a good option, yes?
- I originally put the Ability module in SocialStream::Base::Ability, but since it is included by a model like 'Subject', I thought it was a good idea to put it in 'models'. Besides, if it was to be overloaded by other modules (events, for instance), it looked to me as if it should be something more general.  On the other hand, it only sets authorization rules for social_stream-base entities... so I can change it, I think you have a better notion of how that should work :)
  The problem comes with SocialStream::Ability... I would like to have this module, that would include SocialStream::Base::Ability and could be overloaded by other modules (as events). But if we define this module, the helper methods under SocialStream::Controller::Helpers will look for the class 'Ability'. It traverses its ancestors bottom-up and the first match it will find will be SocialStream::Ability, and not Ability, thus trying to instantiate a module :-s

I'll give it a look to see if I can avoid searching for Ability inside 'SocialStream' module scope.

Maybe we could just have SocialStream::Base::Ability? (btw, do you agree with its location in lib/social_stream/base)?

atd commented:
- About the generator, why is it easier to find by new users? I do not understand this point
- I agree with SocialStream::Ability being a container of all the other Ability modules (base, events, etc..) I have found the key point where the wrong Ability module is loaded:

https://github.com/ging/social_stream/blob/master/base/lib/social_stream/controllers/helpers.rb#L93

Changing it to `::Ability.new` should do the work

carlosgarciaq commented:
Yeap, using ::Ability there works! Thanks!

I've moved the generator to social_stream-base but its name still remains social_stream:add_authorization_rules. I haven't changed its name to social_stream:base:add_authorization_rules because I think it looks to the new user more like internal stuff. And if you are adding rules not related to the 'base' module, I don't know if it makes sense to call a 'base' generator. 
But if you think it's better to name it with base, because it is inside the base module, just tell me, I'll rename it in a second :)

Well, and if you have time and motivation... I was wondering why SocialStream::Controllers::Helpers is trying to access SocialStream::Ability first... (when not written like ::Ability). 
This module is supposed to be included in ActiveSupport (inside base/lib/social_stream/base/engine.rb) and in SocialStream::Controllers::Subject, which in turn is included by Group/UsersController. 
The inclusion of the module should put the methods out of the "SocialStream::Controllers::Helpers" scope, shouldn't it?

atd commented:
Ok, I misread the generator code. I though it was included in the social_stream's base install generator. What do you think about integrating it?

About the last question, I guess it is how name resolution works in Ruby. The method `current_ability` is defined in `SocialStream::Controllers::Helpers`, so Ruby tries to find  `SocialStream::Controllers::Helpers::Ability`, `SocialStream::Controllers::Ability` and finally `SocialStream::Ability`

carlosgarciaq commented:
Yes, it doesn't hurt to have ability.rb from the start... I saw the app directory so clean, that I thought it was a pity to blur it :)

atd commented:
BTW, If you have the interest, if would be awesome if the generator detected if there is already an ability file in the appliication and injected the `include SocialStream::Ability` statement if it already existed

atd commented:
@carlosgarciaq could you please merge all the commits in one, so the pull request keeps clearer?

carlosgarciaq commented:
Hi @atd !
About the generator, if we insert the `include SocialStream::Ability` we should also insert the `super` in the `initialize` method, if it exists (and doesn't already have it)... isn't it?

atd commented:
Yes @carlosgarciaq I think so!

carlosgarciaq commented:
Took some time, but here it is. I changed the install generator to detect if an ability.rb already exists. The code is a bit long and it uses the gem 'ruby_parser' to detect if 'super' is inside 'Ability.initialize'. Hope it's ok. I rebased my commits... I'm pretty new to git, so I hope this is what you meant by merging commits...

carlosgarciaq commented on base/lib/generators/social_stream/base/install_generator.rb:
```diff
@@ -59,4 +59,71 @@ def create_migration_file
+        end
+      end
+      if ability_class
+        include_found = false
```
not used

atd commented:
@carlosgarciaq  I am afraid that the right tool for this is thor https://github.com/wycats/thor, which is already included in Rails' generators http://guides.rubyonrails.org/generators.html 

It has methods to do this things very easily, take a look at http://rdoc.info/github/wycats/thor/master/Thor/Actions

An additional problem with ruby_parser is adding just another dependency to the already bloated list in social_stream-base

carlosgarciaq commented:
Hi @atd,

I use Thor (inject_into_file, create_file, append_to_file) whenever is possible.
The problem is with Thor I cannot know whether an `initialize` method already calls `super`, so I don't know if I have to include the 'super' call or not. With Thor I could detect the existence of `super` but I wouldn't know if it belongs to `initialize` or to another method in the class.

Maybe I overcomplicated it. The generator won't work correctly anyway when the existing `Ability` class includes another module or inherits from a class that has a non-empty `initialize` (because there is no way to call both the socialstream::ability and the original ancestor without modifying them).

so.... I suggest to include always `super` (assume that the existing `Ability` class doesn't call `super`in its `initialize` method) and forget about complicated cases (ability.rb with more than one class, with a class that inherits a class or includes a module, etc.).

atd commented:
Sorry @carlosgarciaq I look at the pull request too quickly

I agree that we can not deal with all the possible cases and I also agree with the solution you suggest. Maybe adding an additional message that warns that the Ability class has been touched and you should take a look to it

cgarciaq committed:
Simplified to not make use of `ruby_parser` gem.

carlosgarciaq commented:
Hi @atd! If you want I can merge the commits again. I do not completely see the benefit of it. Let me know if `git rebase` is the thing to do and if something else is needed.

atd commented:
Thank you @carlosgarciaq Please take a look at http://stackoverflow.com/questions/6102297/how-do-i-clean-up-my-github-fork-so-i-can-make-clean-pull-requests

atd commented:
The reason is that with all the discussion and changes, I got lost and cannot review the whole patch

cgarciaq committed:
Modularization of Ability

atd commented:
Merged in #301
